bpf/process/bpf_generic_lsm.c

```diff
@@ -0,0 +1,195 @@
+	msg = map_lookup_elem(&process_call_heap, &zero);
+	if (!msg)
+		return 0;
+	/* Initialize selector index to 0 */
+	msg->sel.curr = 0;
+#pragma unroll
+	for (i = 0; i < MAX_CONFIGURED_SELECTORS; i++)
+		msg->sel.active[i] = 0;
+	/* Initialize accept field to reject */
+	msg->sel.pass = false;
+	msg->tailcall_index_process = 0;
+	msg->tailcall_index_selector = 0;
+	task = (struct task_struct *)get_current_task();
+	/* Initialize namespaces to apply filters on them */
+	get_namespaces(&msg->ns, task);
+	/* Initialize capabilities to apply filters on them */
+	get_current_subj_caps(&msg->caps, task);
+#ifdef __NS_CHANGES_FILTER
+	msg->sel.match_ns = 0;
+#endif
+#ifdef __CAP_CHANGES_FILTER
+	msg->sel.match_cap = 0;
+#endif
+	msg->idx = get_index(ctx);
+	// setup index and function id
+	config = map_lookup_elem(&config_map, &msg->idx);
+	if (!config)
+		return 0;
+	msg->func_id = config->func_id;
+	msg->retprobe_id = 0;
+	/* Tail call into filters. */
+	tail_call(ctx, &lsm_calls, TAIL_CALL_FILTER);
+	return 0;
+}
+
+#define MAIN "lsm/generic_lsm"
+
+__attribute__((section((MAIN)), used)) int
+generic_lsm_event(struct pt_regs *ctx)
+{
+	return generic_lsm_start_process_filter(ctx);
+}
+
+__attribute__((section("lsm/0"), used)) int
+generic_lsm_setup_event(void *ctx)
+{
+	return generic_process_event_and_setup(
+		ctx, (struct bpf_map_def *)&process_call_heap,
+		(struct bpf_map_def *)&lsm_calls,
+		(struct bpf_map_def *)&config_map,
+		(struct bpf_map_def *)data_heap_ptr);
+}
+
+__attribute__((section("lsm/1"), used)) int
+generic_lsm_process_event(void *ctx)
+{
+	return generic_process_event(ctx,
+				     (struct bpf_map_def *)&process_call_heap,
+				     (struct bpf_map_def *)&lsm_calls,
+				     (struct bpf_map_def *)&config_map,
+				     (struct bpf_map_def *)data_heap_ptr);
+}
+
+__attribute__((section("lsm/2"), used)) int
+generic_lsm_process_filter(void *ctx)
+{
+	int ret;
+
+	ret = generic_process_filter((struct bpf_map_def *)&process_call_heap,
+				     (struct bpf_map_def *)&filter_map);
+	if (ret == PFILTER_CONTINUE)
+		tail_call(ctx, &lsm_calls, TAIL_CALL_FILTER);
+	else if (ret == PFILTER_ACCEPT)
+		tail_call(ctx, &lsm_calls, 0);
+	return PFILTER_REJECT;
+}
+
+__attribute__((section("lsm/3"), used)) int
+generic_lsm_filter_arg(void *ctx)
+{
+	return filter_read_arg(ctx, (struct bpf_map_def *)&process_call_heap,
+			       (struct bpf_map_def *)&filter_map,
+			       (struct bpf_map_def *)&lsm_calls,
+			       (struct bpf_map_def *)&config_map,
+			       true);
+}
+
+__attribute__((section("lsm/4"), used)) int
+generic_lsm_actions(void *ctx)
+{
+	return generic_actions(ctx, &maps);
+}
+
+__attribute__((section("lsm/5"), used)) int
+generic_lsm_output(void *ctx)
+{
+	__u64 id = get_current_pid_tgid();
+	bool *block;
+
+	generic_output(ctx, (struct bpf_map_def *)&process_call_heap, MSG_OP_GENERIC_LSM);
+
+	block = map_lookup_elem(&lsm_block_map, &id);
+	if (!block)
+		return 0;
+
+	map_delete_elem(&lsm_block_map, &id);
+	return -1;
```

Comment:
Do you mean that there are different types of non-zero return values for lsm bpf hooks? And these values affect differently on further execution?
Than, it makes sense to have override action with user-provided return value. We can keep LSMblock action for -EPERM, the way we have SigKill and Signal action.

Reply:
Currently, we have an `Override` action that uses `ArgError` to override a return value.
Could we rese the existing `Override` action in LSM hook?

This would mean that we interpret the `Override` action differently if we are in an kprobe or in lsm: in the former we do bpf_override_return() and in the latter we just return. See: https://github.com/cilium/tetragon/blob/a5c666eee73a0a4412203395f9b0f1c348fde8ca/bpf/process/bpf_generic_kprobe.c#L224.

Maybe we can even re-use the `override_tasks` map that is used in kprobes instead of adding a new one: https://github.com/anfedotoff/tetragon/blob/lsm-sensor/bpf/process/types/basic.h#L126-L133

Reply:
Yes, sure, we could. Do we want to keep an LsmBlock action as an alias for Override action with ArgError = -1?

Reply:
> Do we want to keep an LsmBlock action as an alias for Override action with ArgError = -1?

No strong opinion on that.

Reply:
OK, let's just have an override action! We can add a new action in the future  if it is needed.